core/src/main/scala/org/bykn/bosatsu/TypedExpr.scala

```diff
@@ -233,31 +240,46 @@ object TypedExpr {
     case object NoCall extends SelfCallKind
     case object TailCall extends SelfCallKind
     case object NonTailCall extends SelfCallKind
+    
+    val branchSemigroup: Semigroup[SelfCallKind] =
+      Semigroup.instance(_.merge(_))
+
+    val ifNoCallSemigroup: Semigroup[SelfCallKind] =
+      Semigroup.instance(_.ifNoCallThen(_))
   }
 
-  private def callKind[A](n: Bindable, arity: Int, te: TypedExpr[A]): SelfCallKind =
+  /**
+   * assuming expr is bound to nm, what kind of self call does it contain?
+   */
+  def selfCallKind[A](n: Bindable, te: TypedExpr[A]): SelfCallKind =
```

Comment:
this should be the `def apply` of `SelfCallKind` and possibly moved to a new file if it isn't circular